src/v/raft/replicate_batcher.cc

```diff
@@ -34,13 +35,28 @@ replicate_batcher::replicate_batcher(consensus* ptr, size_t cache_size)
   , _max_batch_size_sem(cache_size)
   , _max_batch_size(cache_size) {}
 
-ss::future<result<replicate_result>> replicate_batcher::replicate(
+replicate_stages replicate_batcher::replicate(
   std::optional<model::term_id> expected_term, model::record_batch_reader&& r) {
-    return do_cache(expected_term, std::move(r)).then([this](item_ptr i) {
-        return _lock.with([this] { return flush(); }).then([i] {
-            return i->_promise.get_future();
-        });
-    });
+    ss::promise<> enqueued;
+    auto enqueued_f = enqueued.get_future();
+    auto f = do_cache(expected_term, std::move(r))
+               .then_wrapped([this, enqueued = std::move(enqueued)](
```

Comment:
`then_wrapped` with `try-catch`, could probably check `f.failed()` instead of the exception handling.